mindmeld/models/nn_utils/classification.py

```diff
@@ -0,0 +1,490 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2015 Cisco Systems, Inc. and others.  All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Base for custom modules that are developed on top of nn layers that can do
+sequence or token classification
+"""
+import json
+import logging
+import os
+import shutil
+import uuid
+from abc import abstractmethod
+from typing import Dict, Union, List
+
+import numpy as np
+from sklearn.metrics import accuracy_score, f1_score
+from tqdm import tqdm
+
+from .helpers import BatchData
+from .helpers import (
+    get_disk_space_of_model,
+    get_num_weights_of_model
+)
+from .input_encoders import InputEncoderFactory
+from .params import get_default_params
+from .._util import _get_module_or_attr
+from ..containers import GloVeEmbeddingsContainer
+from ...core import Bunch
+from ...path import USER_CONFIG_DIR
+
+try:
+    import torch
+    import torch.nn as nn
+
+    nn_module = _get_module_or_attr("torch.nn", "Module")
+except ImportError:
+    nn_module = object
+    pass
+
+SEED = 6174
+LABEL_PAD_TOKEN_IDX = -1  # value set based on default label padding idx in pytorch
+
+logger = logging.getLogger(__name__)
+
+EMBEDDER_TYPE_TO_ALLOWED_TOKENIZER_TYPES = {
+    "glove": ["whitespace-tokenizer", ],
+    "bert": ["huggingface_pretrained-tokenizer", ]
+}
+
+
+class BaseClassification(nn_module):
+    """
+    A base class for sequence and token classifiation using deep neural nets. The trainable examples
+    inputted to the methods of this class are generally in the form of strings or list of strings.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+        self.name = self.__class__.__name__
+        self.params = Bunch()
+        self.params["name"] = self.name
+        self.encoder = None
+
+        self.ready = False  # True when .fit() is called or loaded from a checkpoint, else False
+        self.dirty = False  # True when the model weights aren't saved to disk yet, else False
+
+    def __repr__(self):
+        return f"<{self.name}> ready:{self.ready} dirty:{self.dirty}"
+
+    def get_default_params(self) -> Dict:
+        return get_default_params(self.__class__.__name__)
+
+    def who_am_i(self) -> str:
```

Comment:
Can this be model_status or a little more informative variable name?

Reply:
yes, changed in next push!